ADD 8.0 purchase_update_proposal

bealdav opened the pull request:
## Purchase amended by supplier
![supplier](https://user-images.githubusercontent.com/1853434/143906049-acb58788-806d-4968-959b-24e014761372.png)



## Same purchase to approved by purchaseman
![purchaser](https://user-images.githubusercontent.com/1853434/143906069-8b3ca7d2-861f-45c1-a568-7d97aad73514.png)

gurneyalex commented:
@bealdav  can you check these:

* https://travis-ci.com/github/OCA/purchase-workflow/jobs/469764390#L324-L327
* https://travis-ci.com/github/OCA/purchase-workflow/jobs/469764390#L498-L499
* https://travis-ci.com/github/OCA/purchase-workflow/jobs/469764391#L923-L933

bealdav commented:
Yes @gurneyalex in fact PR should in draft state, sorry. I'll ping you when my tests'll be fixed and extended for a review.

bealdav committed:
ADD module purchase_update_proposal

bealdav commented on .travis.yml:
```diff
@@ -18,11 +18,12 @@ env:
+  # - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="framework_agreement"
+  # - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="purchase_rfq_bid_workflow"
+  # - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="purchase_requisition_auto_rfq"
+  # - TESTS="1" ODOO_REPO="OCA/OCB" INCLUDE="purchase_requisition_multicurrency,purchase_rfq_bid_workflow,purchase_requisition_bid_selection,purchase_requisition_auto_rfq_bid_selection"
```
I commented to make tests quickly

bealdav commented:
Please @lmignon could you have a look here ?

I use it with Order Response.
Instead to apply changes to picking asked by supplier I create a purchase.line.proposal.
Then if accepted, changes are applied to purchase before to impact pickings.

It's also used with supplier which can't send order response as minimal interface to make manually responses.
In my cases delivery dates may be replenished several times, then several are order response are required.

I don't know if you think it could be useful in OCA ? in order response context ? Thanks

bealdav commented:
Please @gurneyalex could you make a review ?

bealdav commented:
cc @thomaspaulb

florian-dacosta commented on purchase_update_proposal/models/purchase_order.py:
```diff
@@ -0,0 +1,212 @@
+        for rec in self:
+            moves = self.env["stock.move"].search(
+                [("purchase_line_id", "in", rec.order_line.ids)]
+            )
```
Easier to include `state in ('cancel', 'done')` in the search and skip the next check? (I mean, just check if `moves` is empty or not)

florian-dacosta commented on purchase_update_proposal/models/purchase_order.py:
```diff
@@ -0,0 +1,212 @@
+                x for x in vals.keys() if x[:9] != "proposal_"]:
+            # The user is not an Odoo purchaser, we must prevent to update other fields
+            logger.info("Fields being writing %s" % vals.keys())
+            raise UserError(
```
I am not sure about that, seems a bit strong...
I mean, I guess there are several possibilites than a none purchaser may edit some fields on purchase...

Maybe with a custom group, maybe using sudo (ok superuser should probably be a purchaser), maybe using the suspend security module (in that case I think it will still raise here)...

Maybe it is possible to check the write access on `purchase.order` of the current user and raise if write access is denied instead the purchase groups ?

bealdav replied:
I made it overridable but not sure it has completed your expectations !?

florian-dacosta replied:
Not really, I think we can do better so there is no need to override this depending on other modules.
Don't you think that check write rights make more sense ?
Something like :` if not self.check_access_rights('write', raise_exception=False)` instead of `if not self._get_purchase_groups()`

bealdav replied:
In fact I have to raise if I have rights for write but not for some fields.
If I have no rights for write it'll raise later in the code. I don't know what you really mean ?

florian-dacosta reviewed (commented):
I did not do a full review, I'd have to take more time and test to fully understand the module...
I'll try to complete the review soon

thomaspaulb commented on purchase_update_proposal/models/purchase_order_line.py:
```diff
@@ -0,0 +1,58 @@
+                "line_id": rec.id,
+                "order_id": rec.order_id.id,
+            }
+            # We only want create proposal line from pu²rchase line
```
typo

thomaspaulb commented on purchase_update_proposal/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,7 @@
+
+Use cases:
+
+* in case of EDI like an Order Response flow (i.e. UBL) which require to deal with update conditions on purchase lines.
```
@bealdav I assume that the EDI flow is your use case; can you explain so that I may understand better how this module could be of use there functionally?

bealdav replied:
> I wonder if it's not possible to use purchase_order_revision for this?

I thought about it. It has few code but it duplicate so many data, that doesn't wish my customer. But I implement a custom revision, but just with an incremental integer, you're right (maybe that I should shared). But the 2 concepts are decoupled.

bealdav replied:
The use case is here:
![Selection_158](https://user-images.githubusercontent.com/1853434/105466701-ed3c4f80-5c94-11eb-9585-d6eb92fb3c75.png)
When I insert order response (Réponse fournisseur UBL) I populated with this module the diff between Order and Order response.
Then my customer check: approve (imply update purchase) or refuse the data of the order response with complete traceability on changes with chatter without duplicate purchase.order

thomaspaulb commented on purchase_update_proposal/models/purchase_order.py:
```diff
@@ -0,0 +1,212 @@
+        if not self._get_purchase_groups() and [
+                x for x in vals.keys() if x[:9] != "proposal_"]:
+            # The user is not an Odoo purchaser, we must prevent to update other fields
+            logger.info("Fields being writing %s" % vals.keys())
```
being written

thomaspaulb commented on purchase_update_proposal/models/purchase_order.py:
```diff
@@ -0,0 +1,212 @@
+    @api.multi
+    def approve_proposal(self):
+        """
+        We need to update purchase lines with approvaled lines.
```
approved

thomaspaulb reviewed (commented):
This is a pretty cool but complex module. It's hard to review so I also didn't get to the end.

I wonder if it's not possible to use [purchase_order_revision](https://odoo-community.org/shop/product/purchase-order-revisions-2133#attr=4482) for this? Then an order response could create a new revision of the same purchase order. Then it might be less technically complex and less likely to clash with other modules or custom code.

bealdav commented:
> I wonder if it's not possible to use purchase_order_revision for this ?

I thought about it. It has few code but it duplicate so many data, that doesn't join my customer wishes. But I implement a custom revision, but just with an incremental integer, you're right (maybe that I should shared). But the 2 concepts are decoupled.

bealdav committed:
FIX oca feeback @florian-dacosta and @thomaspaulb

tonygalmiche reviewed (approved):
LGTM : Functional review only with runbot

bealdav committed:
IMP purch_proposal: add option to skip computing price

FIX proposal: move proposal unlink to _post_process

IMP proposal: copy line instead create, complete tests

FIX proposal: always display price column

bealdav committed:
IMP purch_proposal: make possible to put a line to zero

IMP purch_proposal: update views

IMP purch_proposal: update translation

FIX purch_proposal: view called in button_update_proposal

bealdav committed:
IMP add proposal_display field for rejected case and some filters

bealdav commented on purchase_update_proposal/tests/test_proposal.py:
```diff
@@ -0,0 +1,123 @@
+        order.proposal_ids[0].qty = 99
+        order.submit_proposal()
+        with self.assertRaises(UserError):
+            order.approve_proposal()
```
test fails with last commit

florian-dacosta committed:
[FIX] Make proper cancel when updating a PO from a proposal

Kev-Roche committed:
FIX view name in purchase_order_line

Kev-Roche committed:
[IMP] purchase_update_proposal cancel null quantity

FIX proposal: approve_proposal()

[FIX] proposal: order cancelled only when all lines cancelled

bealdav committed:
IMP proposal: subscribe own proposal group users to purchase at confirmation

IMP proposal: add _hook_for_cancel_process()

bealdav committed:
FIX proposal: prevent to execute self.myfield with multiple records

Kev-Roche committed:
add cancelled status for lines in supplier view

IMP proposal: ergonomy, add ref field

IMP proposal: add oe_read_only to 'approve' button

FIX purch_upd_proposal: cancelled lines stay cancelled after proposal

bealdav committed:
IMP purch_upd_approval: allow to update delay without reset order

FIX move hook_pending_proposal_approval position call

UPD purchase_update_proposal: refresh translation and pictures

bealdav commented:
Thanks

sebastienbeau commented:
/ocabot merge minor

bealdav commented:
Thanks
